src/Compilers/CSharp/Test/Emit3/ExtensionTypeTests.cs

```diff
@@ -40338,4 +40664,4105 @@ public void InstanceField_Readonly()
+  IL_0010:  ldstr      " "
+  IL_0015:  call       "void System.Console.Write(string)"
+  IL_001a:  ldarg.1
+  IL_001b:  call       "void System.Console.Write(int)"
+  IL_0020:  ldarg.0
+  IL_0021:  ldfld      "C E.<UnderlyingInstance>$"
+  IL_0026:  callvirt   "int C.Increment()"
+  IL_002b:  pop
+  IL_002c:  ret
+}
+""");
+    }
+
+    [Fact]
+    public void AdjustReceiver_Invocation_NoHome_ValueType()
+    {
+        var source = $$"""
+new S().M();
+
+struct S { }
+
+implicit extension E for S
+{
+    public void M() { System.Console.Write("ran"); }
+}
+""";
+        var comp = CreateCompilation(source, targetFramework: TargetFramework.Net70);
+        var verifier = CompileAndVerify(comp, expectedOutput: IncludeExpectedOutput("ran"), verify: Verification.FailsPEVerify).VerifyDiagnostics();
+        verifier.VerifyIL("<top-level-statements-entry-point>", """
+{
+  // Code size       21 (0x15)
+  .maxstack  1
+  .locals init (S V_0)
+  IL_0000:  ldloca.s   V_0
+  IL_0002:  initobj    "S"
+  IL_0008:  ldloca.s   V_0
+  IL_000a:  call       "ref E System.Runtime.CompilerServices.Unsafe.As<S, E>(ref S)"
+  IL_000f:  call       "void E.M()"
+  IL_0014:  ret
+}
+""");
+    }
+
+    [Fact]
+    public void AdjustReceiver_Invocation_RefLocal_ReferenceType()
+    {
+        var source = $$"""
+C c = new C();
+ref C c2 = ref c;
+c2.M(c2.Increment());
+System.Console.Write(" ");
+System.Console.Write(c.field);
+
+class C
+{
+    public int field;
+    public int Increment() => field += 1;
+}
+
+implicit extension E for C
+{
+    public void M(int i) { System.Console.Write(this.field); System.Console.Write(" "); System.Console.Write(i); this.Increment(); }
+}
+""";
+        var comp = CreateCompilation(source, targetFramework: TargetFramework.Net70);
+        var verifier = CompileAndVerify(comp, expectedOutput: IncludeExpectedOutput("1 1 2"), verify: Verification.FailsPEVerify).VerifyDiagnostics();
+        verifier.VerifyIL("<top-level-statements-entry-point>", """
+{
+  // Code size       53 (0x35)
+  .maxstack  2
+  .locals init (C V_0, //c
+                C& V_1, //c2
+                C V_2)
+  IL_0000:  newobj     "C..ctor()"
+  IL_0005:  stloc.0
+  IL_0006:  ldloca.s   V_0
+  IL_0008:  stloc.1
+  IL_0009:  ldloc.1
+  IL_000a:  ldind.ref
+  IL_000b:  stloc.2
+  IL_000c:  ldloca.s   V_2
+  IL_000e:  call       "ref E System.Runtime.CompilerServices.Unsafe.As<C, E>(ref C)"
+  IL_0013:  ldloc.1
+  IL_0014:  ldind.ref
+  IL_0015:  callvirt   "int C.Increment()"
+  IL_001a:  call       "void E.M(int)"
+  IL_001f:  ldstr      " "
+  IL_0024:  call       "void System.Console.Write(string)"
+  IL_0029:  ldloc.0
+  IL_002a:  ldfld      "int C.field"
+  IL_002f:  call       "void System.Console.Write(int)"
+  IL_0034:  ret
+}
+""");
+    }
+
+    [Fact]
+    public void AdjustReceiver_Invocation_Local_ReferenceType_ValueModifiedInArgument()
+    {
+        var source = $$"""
+C c = new C();
+c.M(c = null);
```

Comment:
Consider testing something similar for the receiver:

```cs
(c = null).M();
```

Reply:
I have a PROTOTYPE comment for testing extension invocations on a null receiver (should we have a warning, should we generate a null check, or should we let the invocation go through). I'll leave such test for then